### DISCUSSION TITLE
Jupyterlite and entry points, a la, nbconvert

### DISCUSSION THREAD

**Discussion by connorferster** [jupyterlite/jupyterlite#1440] [Q&A] [answered]
As noted from previous discussions, nbconvert cannot be installed into Jupyterlite because it has `tornado` as a dependency. I am experimenting with a "patched" (hacked up) version of nbconvert. I have built a local wheel and am listing the local wheel file in my environment.yml file, like this:

```yaml
...
- pip:
    - nbconvert-7.16.4-py3-none-any.whl
```

This patched nbconvert does install successfully but the File -> Save and Export Notebook As... menu remains empty.

In a cell, I executed the following:

```python
from importlib.metadata import entry_points
entry_points(group="nbconvert.exporters")
```
But only got `[]` as result.

It seems that the Jupyterlite build process (or perhaps just the xeus-kernel?) does not register entry points. Is that correct? Or is there more to the story?

**Comment by jtpio** [comment]
> This patched nbconvert does install successfully but the File -> Save and Export Notebook As... menu remains empty.

This is likely because the in-browser patched nbconvert endpoint returns nothing at the moment:

https://github.com/jupyterlite/jupyterlite/blob/b77efc239521666872535aa9006c4348d8bc6971/packages/server-extension/src/index.ts#L420-L432

> It seems that the Jupyterlite build process (or perhaps just the xeus-kernel?) does not register entry points. Is that correct? Or is there more to the story?

Maybe it could be worth checking with another package that also exposes entry points to make sure the problem does not come from the patched nbconvert, or the packaging of the local wheel?

**Reply by connorferster** [reply]
Ah, I see. Will try out some generic entrypoints and see what happens there.

**Comment by bollwyvl** [comment, marked as answer]
The chain goes deeper, tornado is just the _first_ issue. 

As mentioned elsewhere, I believe the right play is going to end being a port of the core `jinja`-based machinery from nbconvert to a js/ts `nunjucks` in a lite server extension (still mostly lazily loaded). At build time, this would extracting relevant templates and other resources from the well-known locations in the build environment. We have all the pieces for most of the post/preprocessors that are normally handled by a server-like entity, including execution and custom templates provided by contents. This would best be demonstrated first as a separately installable addon.
